tests/graphql_spec.rs

```diff
@@ -23,16 +23,61 @@ mod graphql_mock;
 
 static INIT: Once = Once::new();
 
+#[derive(Debug, Clone, PartialEq)]
+enum Tag {
+  ClientSDL,
+  ServerSDL,
+  MergedSDL,
+}
+
+#[derive(Debug, Clone)]
+struct Source {
+  sdl: String,
+  tag: Tag,
+}
+
 #[derive(Debug, Default, Setters)]
 struct GraphQLSpec {
   path: PathBuf,
-  client_sdl: String,
-  server_sdl: Vec<String>,
-  merged_server_sdl: String,
+  sources: Vec<Source>,
   sdl_errors: Vec<SDLError>,
   test_queries: Vec<GraphQLQuerySpec>,
 }
 
+impl GraphQLSpec {
+  fn client_sdl(mut self, sdl: String) -> Self {
+    self.sources.push(Source { sdl, tag: Tag::ClientSDL });
+    self
+  }
+
+  fn server_sdl(mut self, sdl: Vec<String>) -> Self {
+    for s in sdl {
+      self.sources.push(Source { sdl: s, tag: Tag::ServerSDL });
+    }
+    self
+  }
+
+  fn merged_server_sdl(mut self, sdl: String) -> Self {
+    self.sources.push(Source { sdl, tag: Tag::MergedSDL });
+
+    self
+  }
+
+  fn find_source(&self, tag: Tag) -> String {
+    self.get_sources(tag).first().unwrap().to_string()
+  }
+
+  fn get_sources(&self, tag: Tag) -> Vec<String> {
+    self
+      .sources
+      .clone()
+      .into_iter()
+      .filter(|s| s.tag == tag)
+      .map(|s| s.sdl)
+      .collect()
```

Comment:
it is not a big deal in tests but could be a good pattern to follow especially when writing app's code:

instead of heavily use `clone` and `collect` we can use iterators as is and do any actions only in the end. So here we can return `impl Iterator` and consume it later

Reply:
Regarding `collect`, it makes a lot of sense (I've implemented the change already), but I could not eliminate the clone.

I've tried implementing the Copy trait to avoid cloning it, but I have had no success here. Thanks for the feedback :smile: